### ISSUE TITLE
Add concept of isomorphisms to hints?

### ISSUE THREAD

**Issue by mitchellwrosen** [ndmitchell/hlint#359]
There are many hints that are split into multiple definitions, such as:

```    
- warn: {lhs: isLower a || isUpper a, rhs: isAlpha a}
- warn: {lhs: isUpper a || isLower a, rhs: isAlpha a}
```

It would be nice to be able to declare isomorphic expressions only once, e.g. `x || y = y || x`, `mapM_ = flip forM_`, etc, and have hints automagically exploded into all equivalent versions. Has this idea been explored before?

**Comment by ndmitchell** [comment]
We have some isomorphisms in HLint, specifically around `$` and `.` in hints - see https://github.com/ndmitchell/hlint/blob/master/src/Hint/Match.hs#L21. While you could generalise it, we only have two concrete sides of `||` in one case (the one you mention above). Probably the most pervasive is `<$>` and `fmap`, where we often have hints both ways round, but that's not universal.

I guess I'm not sure that any mechanism would be powerful enough to catch a lot of cases, precise enough not to catch too many, and not more effort than it saves.

**Comment by ChShersh** [comment]
Issues #434 looks related.

> would be powerful enough to catch a lot of cases, precise enough not to catch too many, and not more effort than it saves.

It would be really nice to have only single hint:

```yaml
- warn: {lhs: isLower a || isUpper a, rhs: isAlpha a}
```

Which can suggest to replace `isUpper a || isMyPredicate a || isLower a` with `isAlpha a || isMyPredicate a`. But I guess this is too difficult to implement...
